contributors/design-proposals/security-model.md

```diff
@@ -0,0 +1,190 @@
+resources in a multi-tenant environment.
+
+## Concerns with method based access controls
+
+Many Kubernetes deployments provide method based access control (e.g.,
+list pods). This is akin to securing \*nix with permissions on the "open"
+syscall. To allow more than privileged user, ACLs should be placed on nouns, not
+verbs.
+
+## Concerns with implicit relationships
+
+Kubernetes' implicit object relationships create subtle dangers.
+
+1. The "create pod" privilege allows a user to mount a secret, even if the user
+   cannot "get secrets". This is a failure of confidentiality.
+2. The "create pod" privilege can be used to place a pod behind a service, even
+   if the user cannot "update service". This is a failure of integrity.
+3. The "create pod" privilege can be used to place a pod behind a deployment,
+   which can be used to starve legitimate replicas. This is a failure of
+   integrity, and availability.
+
+This isn't a complete list of such dangers, but it's sufficient to show that
+access controls on an object are insufficient to guarantee CIA. Access controls
+are also needed on the relationships between objects.
+
+## Label ACLs (implicit ownership)
+
+Label ACLs are sometimes suggested as a method to provide access control on
+object-object relationships. Special permissions would be required to create or
+query an ACL'ed label. This approach is both complex and insufficient.
+
+1. The accidental omission of a label means that the system fails open.
+2. It becomes increasingly difficult to reason about permissions, especially if
+   set-based label selectors are used.
+3. Can ACLs be created on existing labels?
+   1. Yes: clients can be locked out of their workloads.
+   2. No: workloads must be restarted to change permissions.
+
+This approach also perturbs the existing behavior of label selectors. Does the
+"list pod" permission allow me to see pods with contentious ACLs? What about
+"list secrets"?
+
+Finally, this approach is insufficent, because not all relationships are defined
+by label selectors. Notably, secrets cannot be protected by label ACLs.
+
+## Explicit "ownership" (recommended)
+
+Explicit owners can be used to limit access to both objects and relationships. A
+service would reject both unauthorized changes to its body, and unauthorized
+pods as backends.
+
+The simplicity and correctness of this method recommend it.
+
+### Namespace as "owner"
+
+By scoping method permissions to namespaces, a crude ownership model can be
+constructed (all objects are owned by a namespace, and object-object
+relationships must have the same owner). The advantage of this approach is that
+it's possible today.
+
+There are also disadvantages to this method.
+
+1. Not all resources are namespaced (e.g., nodes and volumes)
+2. The myriad escalations and CIA failures make it harder to [separate
+   privileges](https://en.wikipedia.org/wiki/Privilege_separation)
+   within a namespace.
```

Comment:
Note that separating privileges within a namespace has never been a core goal of Kubernetes - the original security model explicitly describes namespaces as a security boundary, rather than within a namespace.  Sub-namespace separation should be possible, although I don't know that I think it is required if namespace separation can better model actual user intent.